Add documentation for presentations functionality

MTWhyte opened the pull request.

Commit:
Intermediate progress documenting helper functions

Commit:
Add replace_subword documentation

Commit:
Indentation

Commit:
Add some examples

Commit:
Remove remove_trivial_rules documentation

Commit:
Add a further example

Commit:
Intermediate progress on documenting presentation members

Commit:
Intermediate progress

Commit:
Improve hyperlinking and ToC

Commit:
Remove dosctrings

Commit:
Improve sectioning

Commit:
Delete example documentation

Commit:
Update explanation of sort_rules

Commit:
Improve cross-referencing

Commit:
Make improvements to style

james-d-mitchell commented on src/present.cpp:
```diff
@@ -65,144 +65,39 @@ namespace libsemigroups {
-
-               :Returns: The `Presentation` instance `self`.
-               )pbdoc")
+               py::arg("n"))
```
```suggestion
```

james-d-mitchell commented on src/present.cpp:
```diff
@@ -65,144 +65,39 @@ namespace libsemigroups {
-               )pbdoc")
+               py::overload_cast<>(&Presentation<T>::alphabet, py::const_))
           .def("alphabet",
                py::overload_cast<size_type>(&Presentation<T>::alphabet),
```
```suggestion
               py::overload_cast<size_type>(&Presentation<T>::alphabet))
```

james-d-mitchell commented on src/present.cpp:
```diff
@@ -65,144 +65,39 @@ namespace libsemigroups {
-
-               :Returns: The `Presentation` instance `self`.
-               )pbdoc")
+               py::arg("lphbt"))
```
```suggestion
```

james-d-mitchell commented on src/present.cpp:
```diff
@@ -65,144 +65,39 @@ namespace libsemigroups {
-               )pbdoc")
+               py::arg("n"))
           .def("alphabet",
                py::overload_cast<T const &>(&Presentation<T>::alphabet),
```
```suggestion
               py::overload_cast<T const &>(&Presentation<T>::alphabet))
```

james-d-mitchell commented on src/present.cpp:
```diff
@@ -65,144 +65,39 @@ namespace libsemigroups {
-
-               :return: An int or char.
-               )pbdoc")
+               py::arg("i"))
```
```suggestion
```

james-d-mitchell commented on src/present.cpp:
```diff
@@ -65,144 +65,39 @@ namespace libsemigroups {
-               )pbdoc")
+               &Presentation<T>::alphabet_from_rules)
           .def("letter",
                &Presentation<T>::letter,
```
```suggestion
               &Presentation<T>::letter)
```

james-d-mitchell commented on src/present.cpp:
```diff
@@ -65,144 +65,39 @@ namespace libsemigroups {
-
-               :return: An int.
-               )pbdoc")
+               py::arg("val"))
```
```suggestion
```

james-d-mitchell commented on src/present.cpp:
```diff
@@ -65,144 +65,39 @@ namespace libsemigroups {
-               )pbdoc")
+               py::arg("i"))
           .def("index",
                &Presentation<T>::index,
```
```suggestion
               &Presentation<T>::index)
```

james-d-mitchell commented on src/present.cpp:
```diff
@@ -65,144 +65,39 @@ namespace libsemigroups {
-
-               :Returns: The `Presentation` instance `self`.
-               )pbdoc")
+               py::arg("val"))
```
```suggestion
```

james-d-mitchell commented on src/present.cpp:
```diff
@@ -65,144 +65,39 @@ namespace libsemigroups {
-               )pbdoc")
+                                   py::const_))
           .def("contains_empty_word",
                py::overload_cast<bool>(&Presentation<T>::contains_empty_word),
```
```suggestion
               py::overload_cast<bool>(&Presentation<T>::contains_empty_word))
```

james-d-mitchell commented on src/present.cpp:
```diff
@@ -65,144 +65,39 @@ namespace libsemigroups {
-               :raises RuntimeError:
-                 TODO
-               )pbdoc")
+               py::arg("c"))
```
```suggestion
```

james-d-mitchell commented on src/present.cpp:
```diff
@@ -65,144 +65,39 @@ namespace libsemigroups {
-               )pbdoc")
+                                   py::const_))
           .def("validate_letter",
                &Presentation<T>::validate_letter,
```
```suggestion
               &Presentation<T>::validate_letter)
```

james-d-mitchell committed:
Remove py::arg's from present.cpp

These are only useful (I think) if we use the inline doc, which we're not, so better to remove these.

james-d-mitchell committed:
Update src/present.cpp

Missed this one, doh

james-d-mitchell reviewed (approved):
Just going to see if the CI runs again, and if does I'll merge this. Thanks @MTWhyte !
